balancer-js/src/modules/pools/apr/apr.ts

```diff
@@ -0,0 +1,332 @@
+    const gaugeBalAprBps = Math.round(
+      (boost * 10000 * gaugeBalEmissionsUsd) / gaugeSupplyUsd
+    );
+
+    return gaugeBalAprBps;
+  }
+
+  /**
+   * Some gauges are holding tokens distributed as rewards to LPs.
+   *
+   * @returns APR [bsp] from token rewards.
+   */
+  async rewardsApr(): Promise<number> {
+    // Data resolving
+    const gauge = await this.liquidityGauges.findBy('poolId', this.pool.id);
+    if (!gauge || !gauge.rewardTokens) {
+      return 0;
+    }
+
+    const rewardTokenAddresses = Object.keys(gauge.rewardTokens);
+
+    // Get each tokens rate, extrapolate to a year and convert to USD
+    const rewards = rewardTokenAddresses.map(async (tAddress) => {
+      /* eslint-disable-next-line @typescript-eslint/no-non-null-assertion */
+      const data = gauge!.rewardTokens![tAddress];
+      if (data.period_finish.toNumber() < Date.now() / 1000) {
+        return 0;
+      } else {
+        const yearlyReward = data.rate.mul(86400).mul(365);
+        const price = await this.tokenPrices.find(tAddress);
+        if (price && price.usd) {
+          const meta = await this.tokenMeta.find(tAddress);
+          const decimals = meta?.decimals || 18;
+          const yearlyRewardUsd =
+            parseFloat(formatUnits(yearlyReward, decimals)) *
+            parseFloat(price.usd);
+          return yearlyRewardUsd;
+        } else {
+          throw `No USD price for ${tAddress}`;
+        }
+      }
+    });
+
+    // Get the gauge totalSupplyUsd
+    const bptPriceUsd = await this.bptPrice();
+    const totalSupplyUsd = gauge.totalSupply * bptPriceUsd;
+
+    let total = 0;
+    for await (const reward of Object.values(rewards)) {
+      total += reward / totalSupplyUsd;
+    }
+
+    return Math.round(10000 * total);
+  }
+
+  /**
+   * 80BAL-20WETH pool is accruing protocol revenue.
+   */
+  async protocolApr(): Promise<number> {
+    const veBalPoolId =
+      '0x5c6ee304399dbdb9c8ef030ab642b10820db8f56000200000000000000000014';
+
+    if (this.pool.id != veBalPoolId) {
+      return 0;
+    }
+
+    const revenue = new ProtocolRevenue(this.feeDistributor, this.tokenPrices);
+
+    const { lastWeekBalRevenue, lastWeekBBAUsdRevenue, veBalSupply } =
+      await revenue.data();
+
+    const { totalShares } = this.pool;
+    const totalLiquidity = await this.totalLiquidity();
+    if (!totalLiquidity) {
+      throw 'totalLiquidity for veBal pool missing';
+    }
+    const bptPrice = parseFloat(totalLiquidity) / parseFloat(totalShares);
+
+    const dailyRevenue = (lastWeekBalRevenue + lastWeekBBAUsdRevenue) / 7;
+    const apr = Math.round(
+      (10000 * (365 * dailyRevenue)) / (bptPrice * veBalSupply)
+    );
+
+    return apr;
+  }
+
+  /**
+   * Composes all sources for total pool APR.
+   *
+   * @returns pool APR split [bsp]
+   */
+  async apr(): Promise<AprBreakdown> {
+    const swapFees = await this.swapFees(); // pool snapshot for last 24h fees dependency
```

Comment:
What do you mean "pool snapshot for last 24 hours dependency"? Is this a todo?

Reply:
it's a todo related to 24h fees. from subgraph we get lifetime fees, so it needs a poolSnapshots from previous day to get the last 24h

Reply:
One improvement that's been in the back burner for a while is to more accurately determine the time travel block used in this time travel query that returns the state of the pool 24h ago. We've been using [this notion](https://github.com/balancer-labs/frontend-v2/blob/develop/src/composables/useSnapshots.ts) of "blockTime" to compute the approximate block number, but ideally we'd want to query an auxiliary subgraph that [returns a block number given a timestamp](https://github.com/Xeonus/v3-info/blob/master/src/hooks/useBlocksFromTimestamps.ts#L24).

Reply:
@markusbkoch Right now I'm getting historic fees data from a PoolSnapshots as: currentTotalFees - previousSnapshotFees. Is it a good approach? Or do we better do two queries to subgraph, current block and historic block? (i'd like to avoid that for perf.)

Reply:
A subgraph with historic block numbers is here now and day ago fee is taken based on that